---
title: "Data visualization hands-on session"
author: "Alan Lorenzetti"
date: "8/11/2020"
output:
  html_document:
    toc: true
    toc_depth: 3
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(DT)
library(tidyverse)

ppi = read_tsv("data/ppi.sif", col_names = F)
colnames(ppi) = c("bait", "interaction", "protein")
ppiFunCat = read_tsv("data/ppiFunCat.tsv")
```

# Introduction

In this session we will try to cover a few topics on data visualization.

1. We are going to reproduce a protein-protein interaction network I've made for a paper published recently. For that, we are going to use Cytoscape.
2. We will explore different types of variables and plot them using the R language environment. More specifically we will use RStudio Cloud and a package called `ggplot2`.

# Network visualization using Cytoscape

## Installing Cytoscape and add-ins (or plugins)

* ClusterMaker2
* Color cast
* yFiles Layout Algorithms

## Data structure

### Protein-protein interaction network

This is the structure of a Simple Interaction File (SIF) 

```{r ppiData, echo = F}
datatable(ppi)
```

### Protein information

```{r proteinInfo, echo = F}
datatable(ppiFunCat)
```

## Plotting a protein-protein interaction network

### Context

Is transcription and translation coupled in _Halobacterium salinarum_ ?

Eukaryotes vs. Prokaryotes cellular organization

```{r ppiFig, out.width = "500px", echo = F}
knitr::include_graphics("img/ppiFig.jpg")
```

RPs physically interact with transcription complex components. Diamonds represent RPs; squares represent transcription complex components. Tagged proteins used as bait in the immunoprecipitation experiment are highlighted by a black border. Arrowheads link bait to coimmunoprecipitated proteins. We labeled each of the seven modules obtained by the Newman-Girvan clustering algorithm using a different color [de Lomana et al., 2020].

### Execution

#### Downloading files

* Download the protein-protein interaction [file](https://alanlorenzetti.github.io/dataVisSession2020/data/ppi.sif).

* Download the protein information [file](https://alanlorenzetti.github.io/dataVisSession2020/data/ppiFunCat.tsv).

#### Step by Step

**1.** Import the protein-protein interaction file `ppi.sif`.  
`File -> Import -> Network from file`  

**2.** Load the protein information file.  

  + In the node table panel, click on `Import Table from File`  

**3.** Click on `Style` tab. Let's change the design of our network.

  + Select the `Sample1` preset style.
  + Change the label of nodes by selecting the `label` column.
  + Change the shape of nodes according to the functional `class` of proteins. Rectangles will represent `Transcription` class. Diamonds will represent `Translation` class.
  + Click on `Node` tab and add a border to the nodes of proteins used as baits. Set size `2`.
  + Make the nodes bigger. Select size `35`.
  + Make the edges wider (size `1.5`) and `black`.
  + Add arrowheads (`Target Arrow Shape; Delta`) to the end of edges.
  
**4.** Apply the Newman-Girvan modularity algorithm to find modules of highly interconnected proteins. Use the default parameters.
`Apps -> clusterMaker -> Community Cluster (GLay)`

**5.** Change the color of nodes according to the modules. We will use a plugin called `Color Cast` to make our lives easier. Select `__glayCluster` as the target data column. We'll apply the `Set2` colors.
`Tools -> Color Cast -> Color Cast`

**6.** Remove all the nodes not classified as `Transcription` or `Translation`.

  + Using the `Node Table` panel, order the class column and select all proteins of those classes.
  + Then right-click the selection and then click on `Select nodes from selected rows`.
  + Hide the unselected nodes. `Select -> Nodes -> Hide Unselected Nodes`.
  
**7.** Apply an automatic layout.

  + Select all proteins used as baits (those with borders).
  + Apply yFiles `Hierarchic Layout Selected Nodes`. `Layouts -> Hierarchic Layout Selected Nodes`.
  
**8.** Save your network and export as an image file.

  + `File -> Save`
  + `File -> Export -> Network to Image`

# Data visualization in R using the `ggplot2` package

## Installing `ggplot2` package

```{r install}
#install.packages("ggplot2")
library(ggplot2)
theme_set(theme_bw())

haloExp = read.delim("data/haloExpression.tsv")
```

## Variable types

### Discrete variables

* Counts:  

  + The number of times you cursed this damn virus this week (5)
  + Number of stars in the Milky Way (250 billion or 2.5 x 10^11)
  + Number of genes in a genome (2600)
  + Number of messenger RNA molecules in a bacterial cell (~1800) [Moran et al. 2013]
  + Length of a messenger RNA (1573 nucleotides)

### Continuous variables

* Measures:

  + Height of an individual (162.56 centimeters) 
  + Weight of an individual (89.77 kilograms)

* Ratios:  

  + GC content, i.e., frequency of Guanine or Cytosine bases in a transcript (e.g. 973/1573 = 0.619)

### Categorical variables

* Characteristics:  
  + Color of eyes (black, brown, blue, green, red)
  + Transcript type (protein coding, transporter RNA, ribosomal RNA)

## Data structure

```{r dataStructure, echo=F}
datatable(haloExp)
```

## Syntax

```{r syntax, echo=F, include = F}
knitr::include_graphics("img/ggplotSyntax.png")
```

## Visualizations

#### Histograms

```{r histogram}
ggplot(data = haloExp, mapping = aes(x = length)) +
  geom_histogram()
```

#### Density curves

```{r density}
ggplot(data = haloExp, mapping = aes(x = length)) +
  geom_density()
```

#### Bar plots

```{r bar}
ggplot(data = haloExp, mapping = aes(x = biological_class)) +
  geom_bar() +
  coord_flip()
```

#### Boxplots

```{r boxplot}
ggplot(data = haloExp, mapping = aes(y = GC, x = biological_class)) +
  geom_boxplot() +
  coord_flip()
```

#### Scatter plots

```{r scatter}
ggplot(data = haloExp, mapping = aes(x = mRNA_expression, y = protein_expression)) +
  geom_point()
```

```{r scatterGC}
ggplot(data = haloExp, mapping = aes(x = mRNA_expression, y = protein_expression, color = GC)) +
  geom_point()
```

#### 3D scatter plots

#### Heat maps

```{r heatmap, out.width = "400px", echo = F}
knitr::include_graphics("img/figureSpectronaut.png")
```

## Exercises

* Plot a density curve of the GC variable.
* 

# Resources

* [RStudio IDE cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/rstudio-ide.pdf)
* [`ggplot2` cheatsheet](https://github.com/rstudio/cheatsheets/raw/master/data-visualization-2.1.pdf) 

# References 

DE LOMANA, A. et al. Selective Translation of Low Abundance and Upregulated Transcripts in Halobacterium salinarum. mSystems, v. 5, n. 4, 28 jul. 2020.

MORAN, M. A. et al. Sizing up metatranscriptomics. The ISME Journal, v. 7, n. 2, p. 237â€“243, fev. 2013.